File: 49-group-anagrams/group-anagrams.cpp
class Solution {
public:
    bool isAna(string s1,string s2){
        if(s1.length() != s2.length()){
            return false;
        }
        sort(s1.begin(),s1.end());
        sort(s2.begin(),s2.end());
        return s1==s2;
    }
    vector<vector<string>> groupAnagrams(vector<string>& strs) {


            vector<bool> isused(strs.size(),false);

            vector<vector<string>> result;

            for(int i=0;i<strs.size();i++){

                if(!isused[i]){
                    vector<string> currentgroup;
                    currentgroup.push_back(strs[i]);
                    isused[i] = true;



                     for(int j=i+1;j<strs.size();j++){
                    if(!isused[j] && isAna(strs[i],strs[j])){
                        currentgroup.push_back(strs[j]);
                        isused[j]=true;
                    }


                }

            result.push_back(currentgroup);


                }

           




            }



            return result;

    }
};

File: 49-group-anagrams/group-anagrams.py

# ---------------------- Group Anagrams Logic ----------------------
# sorted(word)
#   - Treats the string as an iterable of characters.
#   - Returns a NEW LIST containing the characters in alphabetical order.
#   - Example:
#         sorted("eat") -> ['a', 'e', 't']
#         sorted("tea") -> ['a', 'e', 't']
#   - Since anagrams contain the same letters, they produce the same sorted list.
#
# tuple(sorted(word))
#   - Converts the sorted list into a tuple.
#   - Tuples are immutable (cannot be changed) and are hashable,
#     so they can be used as dictionary keys.
#   - Lists are mutable and cannot be dictionary keys.
#   - Example:
#         tuple(sorted("eat")) -> ('a', 'e', 't')
#         tuple(sorted("tea")) -> ('a', 'e', 't')
#   - This tuple acts as a unique "signature" for every anagram group.
#
# maps[key].append(word)
#   - 'maps' is a dictionary where:
#         key   = sorted tuple (anagram signature)
#         value = list of words belonging to that anagram group
#   - Every word with the same key is appended to the same list.
#
# Example:
#     "eat" -> ('a','e','t') -> ['eat']
#     "tea" -> ('a','e','t') -> ['eat','tea']
#     "ate" -> ('a','e','t') -> ['eat','tea','ate']
#
# maps.values()
#   - Returns a dict_values object containing all the dictionary values.
#   - It is a dynamic view, not a normal Python list.
#   - Example:
#         dict_values([
#             ['eat','tea','ate'],
#             ['tan','nat'],
#             ['bat']
#         ])
#
# list(maps.values())
#   - Converts the dict_values view into a regular list.
#   - LeetCode expects List[List[str]], so this conversion is necessary.
#   - Final output:
#         [
#             ['eat','tea','ate'],
#             ['tan','nat'],
#             ['bat']
#         ]
# ---------------------------------------------------------------

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:

        maps =  defaultdict(list)

        for word in strs:


            keys = tuple(sorted(word))
            print(keys)
            maps[keys].append(word)


        return list(maps.values())